storage/http_client.go

```diff
@@ -804,191 +803,28 @@ func (c *httpStorageClient) newRangeReaderXML(ctx context.Context, params *newRa
-		// call.
-		if err := ctx.Err(); err != nil {
-			return nil, err
-		}
-		start := params.offset + seen
-		if params.length < 0 && start < 0 {
-			req.Header.Set("Range", fmt.Sprintf("bytes=%d", start))
-		} else if params.length < 0 && start > 0 {
-			req.Header.Set("Range", fmt.Sprintf("bytes=%d-", start))
-		} else if params.length > 0 {
-			// The end character isn't affected by how many bytes we've seen.
-			req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, params.offset+params.length-1))
-		}
-		// We wait to assign conditions here because the generation number can change in between reopen() runs.
-		if err := setConditionsHeaders(req.Header, params.conds); err != nil {
-			return nil, err
-		}
-		// If an object generation is specified, include generation as query string parameters.
-		if params.gen >= 0 {
-			req.URL.RawQuery = fmt.Sprintf("generation=%d", params.gen)
-		}
-
-		var res *http.Response
-		err = run(ctx, func() error {
-			res, err = c.hc.Do(req)
-			if err != nil {
-				return err
-			}
-
-			if res.StatusCode == http.StatusNotFound {
-				res.Body.Close()
-				return ErrObjectNotExist
-			}
-			if res.StatusCode < 200 || res.StatusCode > 299 {
-				body, _ := ioutil.ReadAll(res.Body)
-				res.Body.Close()
-				return &googleapi.Error{
-					Code:   res.StatusCode,
-					Header: res.Header,
-					Body:   string(body),
-				}
-			}
-
-			partialContentNotSatisfied :=
-				!decompressiveTranscoding(res) &&
-					start > 0 && params.length != 0 &&
-					res.StatusCode != http.StatusPartialContent
-
-			if partialContentNotSatisfied {
-				res.Body.Close()
-				return errors.New("storage: partial request not satisfied")
-			}
-
-			// With "Content-Encoding": "gzip" aka decompressive transcoding, GCS serves
-			// back the whole file regardless of the range count passed in as per:
-			//      https://cloud.google.com/storage/docs/transcoding#range,
-			// thus we have to manually move the body forward by seen bytes.
-			if decompressiveTranscoding(res) && seen > 0 {
-				_, _ = io.CopyN(ioutil.Discard, res.Body, seen)
-			}
-
-			// If a generation hasn't been specified, and this is the first response we get, let's record the
-			// generation. In future requests we'll use this generation as a precondition to avoid data races.
-			if params.gen < 0 && res.Header.Get("X-Goog-Generation") != "" {
-				gen64, err := strconv.ParseInt(res.Header.Get("X-Goog-Generation"), 10, 64)
-				if err != nil {
-					return err
-				}
-				params.gen = gen64
-			}
-			return nil
-		}, s.retry, s.idempotent, setRetryHeaderHTTP(nil))
-		if err != nil {
-			return nil, err
-		}
-		return res, nil
-	}
+	reopen := readerReopen(ctx, req.Header, params, s,
+		func() (*http.Response, error) { return c.hc.Do(req) },
+		func() error { return setConditionsHeaders(req.Header, params.conds) },
+		func(gen int64) { req.URL.RawQuery = fmt.Sprintf("generation=%d", params.gen) })
```

Comment:
Looks like the function `gen int64` arg isn't actually needed/used, you can just remove it for clarity (and use `params.gen` directly for JSON as well).